### PR TITLE
Add SlogSink support to funcr

### PR DESCRIPTION
**Pull request by thockin**
See commits for comments.  This unearthed at least 3 bugs.

funcr now passes all but one of the slog tests (that being the timestamp one).

### CODE REVIEW

File: sloghandler.go [hunk 1]
```diff
@@ -130,7 +130,7 @@ func (l *slogHandler) WithGroup(name string) slog.Handler {
 	copy := *l
 	if l.slogSink != nil {
 		copy.slogSink = l.slogSink.WithGroup(name)
-		copy.sink = l.slogSink
+		copy.sink = copy.slogSink
```
Comment (pohly) [inline sloghandler.go]: Do we have a unit test for this?

**Review by pohly — COMMENTED** [review]
Looks good to me, but I've not tried to really umderstand all implementation details in funcr.

File: sloghandler.go (hunk 1 above)
Comment (thockin) [reply sloghandler.go]: Clearly not. It was not clear to me that this field is ever used if slogSink is set - maybe we should set it nil?

Comment (pohly) [reply sloghandler.go]: The other `With*` methods use `sink`, so `WithGroup("foo").WithName("bar")` probably was broken and dropped the group.

I asked about the unit test because I was wondering whether you had observed a failure or merely noticed it when looking at the code.

Comment (thockin) [reply sloghandler.go]: I found it when single-stepping in the debugger. :)

**Comment by thockin** [comment]
Yeah, I don't expect you to dig into funcr too deeply.  Mostly I was interested in your take on the changes to `TestFuncrHandler()` - specifically that the exceptions list is now reduced to just 1.

It feels like that SHOULD be in the funcr subdir, but we also use `testSlog` to test slogSink ?

**Comment by thockin** [comment]
As I stare at slogr_test, I am a little confused - `testSlogSink` seems like a copy of `slogSink` - why?

**Comment by thockin** [comment]
I pushed another commit - please look at the "Get rid of testSlogSink" and tell me if it is wrong?

**Comment by thockin** [comment]
I pushed on more commit, moving slogtests to a helper and funcr tests to funcr.

**Comment by thockin** [comment]
What would you say to having an internal trivial implementation which is basically discard but not nil?  We could have one which supports SlogSink and one which does not.

**Comment by thockin** [comment]
Are we generating a coverage report for PRs?

**Comment by pohly** [comment]
> What would you say to having an internal trivial implementation which is basically discard but not nil?

For `slogtest.TestHandler`, that test sink would have to produce valid JSON. I can suggest a solution tomorrow.

> Are we generating a coverage report for PRs?

No. I was testing locally because I wanted to understand whether moving the testing around had an effect. There's an effect of the PR, it's just because for another reason.

File: benchmark/benchmark_slog_test.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	"log/slog"
+	"os"
+	"testing"
+
+	"github.com/go-logr/logr"
+)
+
+//
+// slogSink wrapper of discard
+//
```
Comment (pohly) [inline benchmark/benchmark_slog_test.go:32]: Did you learn anything from these benchmarks? I'm just curious.

Comment (thockin) [reply benchmark/benchmark_slog_test.go:32]: I learned that funcr with JSON is more than 2x faster than slog's JSON handler, but has more allocs.  And that the slogSink (wrapper) path is (as expected) low overhead.

**Comment by thockin** [comment]
I am cooking something wrt test impls and coverage.

**Comment by thockin** [comment]
I have an extra commit in flight, but opinion wanted.

I can make `slogHandler` do groups properly, but it would store them itself and bypass calling `sink.WithValues()` .  It would come close to passing slog tests, though.  Worthwhile?  Or just leave it with things like `a.b=<value>` ?

**Comment by thockin** [comment]
OK, I added a big commit.  It makes sloghandler pass almost all of the slog tests wrt `WithGroup` and `WithAttrs`, at the cost of bypassing the LogSink.WithValues.  It's not yet safe - maps get copied and then mutated nor have I really looked at performance.  It needs maps in order to get groups right, and that makes me sad (expensive, loses ordering).

I could abandon the maps part and just go back to dot-prefixes (simpler, won't pass as many slog tests but maybe OK?).  Opinion wanted.

File: benchmark/benchmark_slog_test.go (hunk 2 above)
Comment (pohly) [reply benchmark/benchmark_slog_test.go:32]: I did a similar comparison between slog's JSON handler and zap (https://github.com/kubernetes/kubernetes/pull/120696/commits/9100c3cb8a38caf70009f596de10853c473f8db0). Slog was slower, so there was no justification for switching from zap to slog as backend in Kubernetes.

**Comment by pohly** [comment]
> I can make slogHandler do groups properly, but it would store them itself and bypass calling sink.WithValues() . It would come close to passing slog tests, though. Worthwhile? Or just leave it with things like a.b=<value> ?

I prefer to keep `slogHandler` as simple as possible, i.e. use `a.b=<value>` for those sinks which don't implement `SlogSink`. There are other things that sinks should support if they care about being used with slog, like special slog values, so eventually either a sink supports slog and `SlogSink` or it should be replaced with something else. Once we are at that point, the additional complexity in `slogHandler` won't be needed.

**Comment by thockin** [comment]
After having done it, I think I agree.  Though the testing is satisfying :)

**Comment by thockin** [comment]
One last thought: as it is, with prefixes, someone who calls:

```
slogger.Info("msg", slog.Group("g", "k", "v"))
```

gets a different result than:

```
slogger.WithGroup("g", Info("msg", "k", "v"))
```

It's `{"g": { "k": "v"}}` in the first and `{"g.k": "v"}` in the second.

**Comment by thockin** [comment]
I pushed now with a much simpler change to sloghandler, coverage is restored, tests are cleaner (IMO).  PTAL.

File: context_slog_test.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	"log/slog"
+	"os"
+	"testing"
+)
+
+func TestContextWithSlog(t *testing.T) {
+	ctx := context.TODO()
+
+	if out := FromContextAsSlogLogger(ctx); out != nil {
+		t.Errorf("expected error, got %#v", out)
```
Comment (pohly) [inline context_slog_test.go:33]: "expected *no logger*, got ..."

File: context_slog_test.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	// ...read as logr
+	if out, err := FromContext(sctx); err != nil {
+		t.Errorf("unexpected error: %v", err)
+	} else if _, ok := out.sink.(*slogSink); !ok {
+		t.Errorf("expected output to be type *logr.slogSink, got %T", out.sink)
+	}
+
+	// ...read as slog
+	if out := FromContextAsSlogLogger(sctx); out == nil {
+		t.Errorf("unexpected error")
```
Comment (pohly) [inline context_slog_test.go:49]: "did not get any logger"

File: context_test.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+package logr
+
+import (
+	"context"
+	"testing"
+)
+
+func TestContext(t *testing.T) {
+	ctx := context.TODO()
```
Comment (pohly) [inline context_test.go:25]: Background

File: .github/workflows/lint.yaml [hunk 6]
```diff
@@ -11,6 +11,11 @@ jobs:
     steps:
       - name: Checkout code
         uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+      - name: Update Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: '>=1.21.0'
+          cache: false
```
Comment (pohly) [inline .github/workflows/lint.yaml:18]: I agree that we should cover the *new* code with linting. I don't think it's worth to lint also with older Go because the non-slog code isn't that complex and going away at some point.

File: sloghandler.go [hunk 7]
```diff
@@ -70,9 +70,7 @@ func (l *slogHandler) Handle(ctx context.Context, record slog.Record) error {
 
 	kvList := make([]any, 0, 2*record.NumAttrs())
 	record.Attrs(func(attr slog.Attr) bool {
-		if attr.Key != "" {
-			kvList = append(kvList, l.addGroupPrefix(attr.Key), attr.Value.Resolve().Any())
-		}
+		kvList = attrToKVs(attr, l.groupPrefix, kvList)
```
Comment (pohly) [inline sloghandler.go:73]: I think this code change deserves to be a separate commit with a description of why it is made. Right now it is in "Clean up slog testing and restore coverage" which doesn't tell me why sloghandler.go needs to be changed.

File: sloghandler.go [hunk 8]
```diff
@@ -127,21 +123,50 @@ func (l *slogHandler) WithGroup(name string) slog.Handler {
 	if l.sink == nil {
 		return l
 	}
+	if name == "" {
+		// slog says to inline empty groups
+		return l
+	}
```
Comment (pohly) [inline sloghandler.go:129]: This is a bug fix, right?

A separate PR would make it show up in the 1.4 release notes.

File: context_slog_test.go (hunk 3 above)
Comment (thockin) [reply context_slog_test.go:33]: ACK

File: context_slog_test.go (hunk 4 above)
Comment (thockin) [reply context_slog_test.go:49]: ACK

File: context_test.go (hunk 5 above)
Comment (thockin) [reply context_test.go:25]: ACK

File: .github/workflows/lint.yaml (hunk 6 above)
Comment (thockin) [reply .github/workflows/lint.yaml:18]: i cant tell if you are agreeing with me or want a change?

**Comment by thockin** [comment]
pushed with small changes, will look at splitting sloghandler to a new PR or beefing up comments.

File: sloghandler.go (hunk 8 above)
Comment (thockin) [reply sloghandler.go:129]: https://github.com/go-logr/logr/pull/245

File: sloghandler.go (hunk 7 above)
Comment (thockin) [reply sloghandler.go:73]: I can either a) break this last commit to a PR which applies AFTER this PR (rebasing it is tedious and it has a lot of deps on other commits); or b) beef up the comments.

I'll do B now and you can decide if you prefer A.

File: sloghandler.go [hunk 9]
```diff
@@ -127,21 +123,50 @@ func (l *slogHandler) WithGroup(name string) slog.Handler {
...
 	return &clone
 }
 
-func (l *slogHandler) addGroupPrefix(name string) string {
-	if l.groupPrefix == "" {
+// attrToKVs appends a slog.Attr to a logr-style kvList.  It handle slog Groups
+// and other details of slog.
+func attrToKVs(attr slog.Attr, groupPrefix string, kvList []any) []any {
+	attrVal := attr.Value.Resolve()
+	if attrVal.Kind() == slog.KindGroup {
```
Comment (pohly) [inline sloghandler.go:144]: I suppose this is the "retools sloghandler to handle slog Attrs a bit better" part.

Looks okay to me now. Let's rebase and then merge it without further refactoring (your plan B).

File: .github/workflows/lint.yaml (hunk 6 above)
Comment (pohly) [reply .github/workflows/lint.yaml:18]: I agree with you.
